### PR TITLE
Cached chunk

### PR DESCRIPTION
**Pull request by JackieTien97**

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/cache/ChunkCache.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+    cacheRequestNum.incrementAndGet();
+
+    try {
+      readLock.lock();
+      if (lruCache.containsKey(chunkMetaData)) {
+        cacheHitNum.incrementAndGet();
+        printCacheLog(true);
+        Chunk chunk = lruCache.get(chunkMetaData);
+        return new Chunk(chunk.getHeader(), chunk.getData().duplicate(), chunk.getDeletedAt(), reader.getEndianType());
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/cache/ChunkCache.java]: why not return the chunk got from cache?

File: server/src/main/java/org/apache/iotdb/db/engine/cache/ChunkCache.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  public static ChunkCache getInstance() {
+    return ChunkCacheHolder.INSTANCE;
+  }
+
+  public Chunk get(ChunkMetaData chunkMetaData, TsFileSequenceReader reader) throws IOException {
+    if (!cacheEnable) {
+      Chunk chunk = reader.readMemChunk(chunkMetaData);
+      return new Chunk(chunk.getHeader(), chunk.getData().duplicate(), chunk.getDeletedAt(), reader.getEndianType());
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/cache/ChunkCache.java]: why new a Chunk?

File: server/src/main/java/org/apache/iotdb/db/engine/cache/ChunkCache.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      printCacheLog(true);
+      Chunk chunk = lruCache.get(chunkMetaData);
+      readLock.unlock();
+      return new Chunk(chunk.getHeader(), chunk.getData().duplicate(), chunk.getDeletedAt(), reader.getEndianType());
+    }
+    printCacheLog(false);
+    Chunk chunk = reader.readMemChunk(chunkMetaData);
+    lruCache.put(chunkMetaData, chunk);
+    writeLock.unlock();
+    return new Chunk(chunk.getHeader(), chunk.getData().duplicate(), chunk.getDeletedAt(), reader.getEndianType());
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/cache/ChunkCache.java]:
```suggestion
    return chunk;
```

File: server/src/main/java/org/apache/iotdb/db/engine/cache/ChunkCache.java (hunk 1 above)
Comment (JackieTien97) [reply server/src/main/java/org/apache/iotdb/db/engine/cache/ChunkCache.java]: The chunk has a field byte buffer, it should be reused.

File: server/src/main/java/org/apache/iotdb/db/engine/cache/ChunkCache.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  private final ReadWriteLock lock = new ReentrantReadWriteLock();
+  private final Lock readLock = lock.readLock();
+  private final Lock writeLock = lock.writeLock();
+
+
+  private ChunkCache() {
+    lruCache = new LRULinkedHashMap<ChunkMetaData, Chunk>(MEMORY_THRESHOLD_IN_CHUNK_CACHE, true) {
+      @Override
+      protected long calEntrySize(ChunkMetaData key, Chunk value) {
+        return RamUsageEstimator.sizeOf(value) + RamUsageEstimator.sizeOf(value);
```
Comment (liutaohua) [inline server/src/main/java/org/apache/iotdb/db/engine/cache/ChunkCache.java]: why this return valueSize * 2 ?

File: server/src/main/java/org/apache/iotdb/db/engine/cache/ChunkCache.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        Chunk chunk = lruCache.get(chunkMetaData);
+        return new Chunk(chunk.getHeader(), chunk.getData().duplicate(), chunk.getDeletedAt(), reader.getEndianType());
+      }
+    } finally {
+      readLock.unlock();
+    }
+
+    writeLock.lock();
+    if (lruCache.containsKey(chunkMetaData)) {
+      writeLock.unlock();
```
Comment (liutaohua) [inline server/src/main/java/org/apache/iotdb/db/engine/cache/ChunkCache.java]: Why not down the lock to readLock

File: server/src/main/java/org/apache/iotdb/db/query/reader/chunk/DiskChunkLoader.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  @Override
+  public void close() throws IOException {
+    reader.close();
+  }
+
+  @Override
+  public void clear() {
+    // no cache need clear
```
Comment (liutaohua) [inline server/src/main/java/org/apache/iotdb/db/query/reader/chunk/DiskChunkLoader.java]: i think , this method can delete for  interface, neither memChunk nor diskChunk don't need to clear ,and the cache is  LRU

File: server/src/main/java/org/apache/iotdb/db/engine/cache/ChunkCache.java (hunk 4 above)
Comment (JackieTien97) [reply server/src/main/java/org/apache/iotdb/db/engine/cache/ChunkCache.java]: it's a typo, should be RamUsageEstimator.sizeOf(key) + RamUsageEstimator.sizeOf(value);
Fixed

**Pull request merged**